source/Server/IntegratedAuthentication/IntegratedChallengeCoordinator.cs

```diff
@@ -64,10 +66,29 @@ public IntegratedChallengeTrackerStatus SetupResponseIfChallengeHasNotSucceededY
                         Secure = state?.UsingSecureConnection ?? false
                     });
                 
-                // we pass back to the original link here (e.g. the deep link that originally triggered the sign in)
+                // we pass back to the original link here if it is a trusted url (e.g. the deep link that originally triggered the sign in)
                 // we normally wouldn't do this without the user being authenticated, but given we know they aren't
                 // authenticated we'll redirect back and rely on the sign in page kicking in again and seeing the cookie.
-                context.Response.Redirect(stateRedirectAfterLoginTo);
+                // If the caller has provided a redirect after successful login, we need to check it is a local address - preventing Open Redirection attacks
+                if (!string.IsNullOrWhiteSpace(state?.RedirectAfterLoginTo))
+                {
+                    var whitelist = authenticationConfigurationStore.GetTrustedRedirectUrls();
+                    if (Requests.IsLocalUrl(state.RedirectAfterLoginTo, whitelist))
+                    {
+                        // This is a safe redirect, let's go!
+                        context.Response.Redirect(state.RedirectAfterLoginTo);
+
+                        return IntegratedChallengeTrackerStatus.ChallengeFailed;
```

Comment:
We're returning right away without calling `SetChallengeCompleted` like the old flow. Is this intended?

Reply:
Nope, good catch 👍

Reply:
Cool. I have a follow up suggestion since now the code paths are very similar.

Wonder if we can use a variable `var redirectUrl = context.Request.PathBase.Value ?? "/";` as the default. Then your conditional block only needs to worry about updating the `redirectUrl` when it's safe or logging a warning when it's not.

Everything can converge onto
```csharp
context.Response.Redirect(redirectUrl);
SetChallengeCompleted(context.Connection.Id);
return IntegratedChallengeTrackerStatus.ChallengeFailed;
```

A little bit DRYer.

Reply:
That's a great idea. Updated